File: app/src/main/java/com/bakkenbaeck/sol/ui/InfoFragment.java
package com.bakkenbaeck.sol.ui;

import android.databinding.DataBindingUtil;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v4.content.ContextCompat;
import android.text.Html;
import android.text.Spanned;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.bakkenbaeck.sol.R;
import com.bakkenbaeck.sol.databinding.FragmentInfoBinding;
import com.bakkenbaeck.sol.util.CurrentPhase;
import com.bakkenbaeck.sol.util.SolPreferences;

public class InfoFragment extends Fragment {
    public static final String TAG = "InfoFragment";
    public static final String PHASE_NAME = "InfoFragment";

    private FragmentInfoBinding binding;

    private boolean notifications;
    private String phaseName;

    public static InfoFragment newInstance(final String phaseName) {
        Log.d(TAG, "newInstance: " + phaseName);
        Bundle b = new Bundle();
        b.putString(PHASE_NAME, phaseName);
        InfoFragment f = new InfoFragment();
        f.setArguments(b);
        return f;
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        binding = DataBindingUtil.inflate(inflater, R.layout.fragment_info, null, false);

        init();

        return binding.getRoot();
    }

    private void init() {
        if (this.phaseName == null) {
            this.phaseName = getArguments().getString(PHASE_NAME);
        }

        setColors(phaseName);

        this.binding.titleWrapper.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                ((SunActivity) getActivity()).onBackPressed();
            }
        });

        SolPreferences solPrefs = new SolPreferences(getContext());
        notifications = solPrefs.getShowNotification();

        String s = notifications
                ? "off"
                : "on";

        this.binding.notificationValue.setText(s);

        this.binding.notificationWrapper.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                notifications = !notifications;

                String s = notifications
                        ? "off"
                        : "on";

                binding.notificationValue.setText(s);
                SolPreferences solPrefs = new SolPreferences(getContext());
                solPrefs.cacheShowNotification(notifications);
            }
        });
    }

    private void setColors(final String phaseName) {
        final CurrentPhase phase = new CurrentPhase(phaseName);
        final String infoMessage = this.getContext().getResources().getString(R.string.info_message);
        final String formatedInfo = infoMessage.replace("{color}", String.valueOf(ContextCompat
                .getColor(this.getContext(), phase.getPrimaryColor())));
        final Spanned s = convertToHtml(formatedInfo);
        this.binding.infoMessage.setText(s);

        final int color = phase.getBackgroundColor();
        final int priColor = phase.getPrimaryColor();
        final int secColor = phase.getSecondaryColor();

        this.binding.title.setTextColor(ContextCompat.getColor(this.getContext(), secColor));
        this.binding.root.setBackgroundColor(ContextCompat.getColor(this.getContext(), color));
        this.binding.infoMessage.setTextColor(ContextCompat.getColor(this.getContext(), secColor));
        this.binding.notificationText.setTextColor(ContextCompat.getColor(this.getContext(), secColor));
        this.binding.notificationValue.setTextColor(ContextCompat.getColor(this.getContext(), priColor));

        setSunDrawable(secColor);
    }

    public void setSunDrawable(final int color) {
        switch (color) {
            case R.color.sunrise_text: {
                this.binding.sunCircle.setImageDrawable(ContextCompat.getDrawable(this.getContext(), R.drawable.half_circle_sunrise));
                break;
            }
            case R.color.daylight_text: {
                this.binding.sunCircle.setImageDrawable(ContextCompat.getDrawable(this.getContext(), R.drawable.half_circle_daylight));
                break;
            }
            case R.color.sunset_text: {
                this.binding.sunCircle.setImageDrawable(ContextCompat.getDrawable(this.getContext(), R.drawable.half_circle_daylight));
                break;
            }
            case R.color.twilight_text: {
                this.binding.sunCircle.setImageDrawable(ContextCompat.getDrawable(this.getContext(), R.drawable.half_circle_twilight));
                break;
            }
            case R.color.night_text: {
                this.binding.sunCircle.setImageDrawable(ContextCompat.getDrawable(this.getContext(), R.drawable.half_circle_night));
                break;
            }
        }
    }

    public void update(final String phaseName) {
        if (this.binding == null) {
            return;
        }

        this.phaseName = phaseName;
        setColors(phaseName);
    }

    private Spanned convertToHtml(final String message) {
        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.N) {
            return Html.fromHtml(message, Html.FROM_HTML_MODE_LEGACY);
        } else {
            return Html.fromHtml(message);
        }
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();

        phaseName = null;

        if (binding != null) {
            binding.unbind();
            binding = null;
        }
    }
}

File: app/src/main/java/com/bakkenbaeck/sol/view/fragment/InfoFragment.kt
package com.bakkenbaeck.sol.view.fragment

import android.animation.ArgbEvaluator
import android.animation.ValueAnimator
import android.graphics.drawable.ColorDrawable
import android.os.Bundle
import android.text.Spanned
import android.text.method.LinkMovementMethod
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.core.content.ContextCompat
import androidx.fragment.app.Fragment
import androidx.fragment.app.viewModels
import androidx.lifecycle.Observer
import androidx.navigation.fragment.findNavController
import com.bakkenbaeck.sol.R
import com.bakkenbaeck.sol.extension.getColor
import com.bakkenbaeck.sol.extension.setTextToOnOrOff
import com.bakkenbaeck.sol.extension.toHtml
import com.bakkenbaeck.sol.model.local.Phase
import com.bakkenbaeck.sol.viewModel.InfoViewModel
import kotlinx.android.synthetic.main.fragment_info.*

class InfoFragment : Fragment() {

    companion object {
        private const val PHASE_NAME = "phaseName"

        private val color2drawable = mapOf(
            R.color.sunrise_text to R.drawable.half_circle_sunrise,
            R.color.daylight_text to R.drawable.half_circle_daylight,
            R.color.sunset_text to R.drawable.half_circle_daylight,
            R.color.twilight_text to R.drawable.half_circle_twilight,
            R.color.night_text to R.drawable.half_circle_night
        )
    }

    private val viewModel: InfoViewModel by viewModels()

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        return inflater.inflate(R.layout.fragment_info, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        init()
    }

    private fun init() {
        initBackground()
        initClickListeners()
        initObservers()
    }

    private fun initBackground() {
        viewModel.animateBackground.value = false
    }

    private fun initObservers() {
        viewModel.notifications.observe(viewLifecycleOwner, Observer {
            if (it != null) notificationValue.setTextToOnOrOff(it)
        })
        viewModel.animateBackground.observe(viewLifecycleOwner, Observer {
            if (it != null) setColorsFromPhaseName(it)
        })
    }

    private fun setColorsFromPhaseName(animateBackground: Boolean = false) {
        val phaseName = arguments?.getString(PHASE_NAME) ?: return
        val currentPhase = Phase(phaseName)

        val primaryColor = getColor(currentPhase.primaryColor)
        val secondaryColor = getColor(currentPhase.secondaryColor)

        infoMessage.movementMethod = LinkMovementMethod.getInstance()
        infoMessage.text = createInfoMessage(primaryColor)

        title?.setTextColor(secondaryColor)
        infoMessage.setTextColor(secondaryColor)
        notificationText.setTextColor(secondaryColor)
        notificationValue.setTextColor(primaryColor)

        initBackground(currentPhase, animateBackground)
    }

    private fun createInfoMessage(primaryColor: Int): Spanned {
        val message = resources.getString(R.string.info_message)
        val formattedInfo = message.replace("{color}", primaryColor.toString())
        return formattedInfo.toHtml()
    }

    private fun initBackground(currentPhase: Phase, animate: Boolean) {
        val colorFrom = (root.background as ColorDrawable).color
        val colorTo = getColor(currentPhase.backgroundColor)
        if (animate) {
            animateBackground(colorFrom, colorTo)
        } else {
            root.setBackgroundColor(colorTo)
        }
        setSunDrawable(currentPhase)
    }

    private fun animateBackground(colorFrom: Int, colorTo: Int) {
        if (colorFrom == colorTo) return

        ValueAnimator.ofObject(ArgbEvaluator(), colorFrom, colorTo).apply {
            duration = 400
            addUpdateListener { root.setBackgroundColor(it.animatedValue as Int) }
            start()
        }
    }

    private fun setSunDrawable(currentPhase: Phase) {
        val color = currentPhase.secondaryColor
        val drawable = color2drawable[color] ?: return
        setSunDrawable(drawable)
    }

    private fun setSunDrawable(resourceId: Int) {
        val context = context ?: return
        val drawable = ContextCompat.getDrawable(context, resourceId)
        sunCircle.setImageDrawable(drawable)
    }

    private fun initClickListeners() {
        notificationWrapper.setOnClickListener { viewModel.toggleNotifications() }
        titleWrapper.setOnClickListener {
            val controller = findNavController()
            controller.navigateUp()
        }
    }
}